# SI 618: Data Manipulation and Analysis
## 03 - Pandas 2
### Dr. Chris Teplovs, School of Information, University of Michigan
<small><a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a> This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.
    
Version 2021.02.04.1.CT

### IMPORTANT: Replace ```?``` in the following code with your uniqname.

In [1]:
MY_UNIQNAME = 'cyuntao'

# Loading and manipulating data in pandas

## Learning Objectives
* load CSV files
* load JSON files
* use pd.read_html to extract tables from web pages
* handle missing data (dropna and fillna)
* use vectorized string functions
* use Pandas' apply function to run a function on each row of a dataframe
* view and set the indexes of a dataframe, including hierarchical indexes
* use loc to explore a dataframe with hierarchical indexes
* use stack and unstack to reshape dataframes
* concatenate two DataFrames by columns
* use Pandas' merge function to join dataframes in a SQL-like way
* use the .describe() function
* understand .groupby()
* know how to use pivot and pivot_table

In [2]:
import pandas as pd
import numpy as np

Recall the ```pd.read_csv``` function that we used to load data sets in previous classes:

In [3]:
menu = pd.read_csv('https://github.com/umsi-data-science/data/raw/main/titles.csv') 

That works great for well-formatted CSV files, but what happens when you get something that looks like the ```data/avocado_eu.csv``` file.

If you're using JupyterLab, go ahead and browse that in JupyterLab's CSV browser.  You'll notice a new drop-down menu labelled "Delimiter".  Go ahead and change that to ```;```.

Referring back to your readings and the [read_csv documentation online](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html), complete the following exercise

Read the data/avocado_eu.csv file into a pandas DataFrame and show the first 5 rows.


In [4]:
avocado = pd.read_csv('https://github.com/umsi-data-science/data/raw/main/avocado_eu.csv')
avocado.head(5)

;Date;AveragePrice;Total Volume;4046;4225;4770;Total Bags;Small Bags;Large Bags;XLarge Bags;type;year;region
0;2015-12-27;1 33;64236  62;1036 74;54454 85;48  16;8696 87;8603 62;93  25;0                         0;conventional;2015;Albany                                                          
1;2015-12-20;1 35;54876  98;674  28;44638 81;58  33;9505 56;9408 07;97  49;0                         0;conventional;2015;Albany                                                          
2;2015-12-13;0 93;118220 22;794  7;109149 67;130 5;8145  35;8042 21;103 14;0                         0;conventional;2015;Albany                                                          
3;2015-12-06;1 08;78992  15;1132 0;71976  41;72  58;5811 16;5677 4;133  76;0                         0;conventional;2015;Albany                                                          
4;2015-11-29;1 28;51039  6;941   48;43838 39;75  78;6183 95;5986 26;197 69;0                         0;conventional;2015;Albany

You'll notice that, unless you did something special in the previous read_csv invocation, the decimal points don't look quite right.  Go ahead and find the right option to convert commas to periods when loading a CSV file.

## <font color="magenta">Q1 (2 points): 
Read the data/avocado_eu.csv file into a DataFrame called "avocado" using the correct delimiter and decimal character into a dataframe and show the first 5 rows: </font>

In [5]:
avocado = pd.read_csv('https://raw.githubusercontent.com/umsi-data-science/si370/master/data/avocado_eu.csv', sep=';', decimal=',')
avocado.head(5)

Unnamed: 0        Date  AveragePrice  Total Volume     4046       4225  \
0          0  2015-12-27          1.33      64236.62  1036.74   54454.85   
1          1  2015-12-20          1.35      54876.98   674.28   44638.81   
2          2  2015-12-13          0.93     118220.22   794.70  109149.67   
3          3  2015-12-06          1.08      78992.15  1132.00   71976.41   
4          4  2015-11-29          1.28      51039.60   941.48   43838.39   

     4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  \
0   48.16     8696.87     8603.62       93.25          0.0  conventional   
1   58.33     9505.56     9408.07       97.49          0.0  conventional   
2  130.50     8145.35     8042.21      103.14          0.0  conventional   
3   72.58     5811.16     5677.40      133.76          0.0  conventional   
4   75.78     6183.95     5986.26      197.69          0.0  conventional   

   year  region  
0  2015  Albany  
1  2015  Albany  
2  2015  Albany  
3  2015  Albany  
4  2015  Albany

## Counting the number of values (review)

Sometimes, you'll want to count the number of times values occur.  For example, we might want to know the number of times each 'type'
is reported in our avocado data.  Use the ```value_counts()``` function on a Series to do so:

In [6]:
avocado['type'].value_counts()

conventional    9126
organic         9123
Name: type, dtype: int64

## Loading JSON data

In addition to CSV files, JSON (JavaScript Object Notation) files or data is commonly used.  

In [7]:
nfl_football_players = pd.read_json('https://github.com/umsi-data-science/data/raw/main/nfl_football_profiles.json')

In [8]:
nfl_football_players.head()

player_id             name position height  weight    current_team  \
0       1809  Robert Blackmon       DB    6-0   208.0            None   
1      23586       Dean Wells       LB    6-3   248.0            None   
2        355      Kiko Alonso      ILB    6-3   238.0  Miami Dolphins   
3      18182     Steve Ramsey       QB    6-2   210.0            None   
4      16250       Cory Nelms       CB    6-0   195.0            None   

   birth_date     birth_place  death_date      college       high_school  \
0  1967-05-12    Bay City, TX        None       Baylor     Van Vleck, TX   
1  1970-07-20  Louisville, KY        None     Kentucky    Holy Cross, KY   
2  1990-08-14      Newton, MA        None       Oregon     Los Gatos, CA   
3  1948-04-22      Dallas, TX  1999-10-15  North Texas  W.W. Samuell, TX   
4  1988-02-27     Neptune, NJ        None   Miami (FL)       Neptune, NJ   

           draft_team  draft_round  draft_position  draft_year current_salary  \
0    Seattle Seahawks          2.0            34.0      1990.0           None   
1    Seattle Seahawks          4.0            85.0      1993.0           None   
2       Buffalo Bills          2.0            46.0      2013.0      1,075,000   
3  New Orleans Saints          5.0           126.0      1970.0           None   
4                None          NaN             NaN         NaN           None   

   hof_induction_year  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN

And, just for fun, show the player with the highest Current Salary from that dataset:

In [9]:
nfl_football_players.sort_values('current_salary', ascending=False).head(1)

player_id               name position height  weight  \
6454        721  Jeremiah Attaochu      OLB    6-3   252.0   

              current_team  birth_date      birth_place death_date  \
6454  Los Angeles Chargers  1993-01-17  Ibadan, Nigeria       None   

           college             high_school          draft_team  draft_round  \
6454  Georgia Tech  Archbishop Carroll, DE  San Diego Chargers          2.0   

      draft_position  draft_year current_salary  hof_induction_year  
6454            50.0      2014.0        993,150                 NaN

## Fixing up the data
Assuming you did something like sort_values on one of the original columns, you probably got the wrong result.

Looking a bit more closely at the results, you'll notice that the current_salary column.  Remembering that we have made the shift from pythonic to pandorable, we can leverage the impressive-sounding "vectorized string functions" mentioned in Section 7.3 of the McKinney book.  Specifically, we can use the str.replace(...) method.  Note that had we use read_csv to load the file we could have used the ```thousands=``` option and avoided all this, but sometimes data doesn't come in a convenient format.

One way to apply functions is to operate on a column and then assign the results to another column.  For example, if we wanted to eliminate commas, we could replace them with null strings

In [10]:
nfl_football_players['current_salary'].str.replace(',', '')

0           None
1           None
2        1075000
3           None
4           None
          ...   
25038       None
25039       None
25040       None
25041       None
25042       None
Name: current_salary, Length: 25043, dtype: object

And assign the results to a column in the original dataframe (in this case I'm calling the column current_salary_nocommas). We need to assign it to a column to ensure the operation is done in-place.

In [11]:
nfl_football_players['current_salary_nocommas'] = nfl_football_players['current_salary'].str.replace(',', '')

But you'll notice that the type of the column is string, and we want to convert it to a float so we can sort it numerically.  So we can use the astype() function to convert it:

In [12]:
nfl_football_players['current_salary_cleaned'] = nfl_football_players['current_salary_nocommas'].astype(float)

In [13]:
nfl_football_players.head(2)

player_id             name position height  weight current_team  \
0       1809  Robert Blackmon       DB    6-0   208.0         None   
1      23586       Dean Wells       LB    6-3   248.0         None   

   birth_date     birth_place death_date   college     high_school  \
0  1967-05-12    Bay City, TX       None    Baylor   Van Vleck, TX   
1  1970-07-20  Louisville, KY       None  Kentucky  Holy Cross, KY   

         draft_team  draft_round  draft_position  draft_year current_salary  \
0  Seattle Seahawks          2.0            34.0      1990.0           None   
1  Seattle Seahawks          4.0            85.0      1993.0           None   

   hof_induction_year current_salary_nocommas  current_salary_cleaned  
0                 NaN                    None                     NaN  
1                 NaN                    None                     NaN

And now we can re-run our command to sort by salary and get the correct result:

In [14]:
nfl_football_players.sort_values('current_salary_cleaned', ascending=False).head(1)

player_id          name position height  weight         current_team  \
17756       4644  Kirk Cousins       QB    6-3   214.0  Washington Redskins   

       birth_date  birth_place death_date       college  \
17756  1988-08-19  Holland, MI       None  Michigan St.   

                 high_school           draft_team  draft_round  \
17756  Holland Christian, MI  Washington Redskins          4.0   

       draft_position  draft_year current_salary  hof_induction_year  \
17756           102.0      2012.0     23,943,600                 NaN   

      current_salary_nocommas  current_salary_cleaned  
17756                23943600              23943600.0

## Dropping missing values

In addition to the "all" or "any" functionality described in McKinney section 7.1, it's sometimes useful to drop a row only if a certain column or columns have missing data.  To do this, use the subset= option with dropna().  So, for example, to drop all players for whom we do not have salary information, we could use the following code:

In [15]:
nfl_football_players.current_salary_cleaned.isna().sum()

23278

In [16]:
nfl_football_players_salaries = nfl_football_players.dropna(subset=['current_salary_cleaned'])

In [17]:
nfl_football_players_salaries.head()

player_id               name position height  weight      current_team  \
2         355        Kiko Alonso      ILB    6-3   238.0    Miami Dolphins   
6        2701      Preston Brown      ILB    6-1   251.0     Buffalo Bills   
13       3966        Frank Clark       DE    6-2   270.0  Seattle Seahawks   
37      19449  Emmanuel Sanders        WR   5-11   186.0    Denver Broncos   
53      17858       Bilal Powell       RB   5-11   204.0     New York Jets   

    birth_date     birth_place death_date     college      high_school  \
2   1990-08-14      Newton, MA       None      Oregon    Los Gatos, CA   
6   1992-10-27  Cincinnati, OH       None  Louisville    Northwest, OH   
13  1993-06-14   Cleveland, OH       None    Michigan    Glenville, OH   
37  1987-03-16   Bellville, TX       None         SMU    Bellville, TX   
53  1988-10-27    Lakeland, FL       None  Louisville  Lake Gibson, FL   

             draft_team  draft_round  draft_position  draft_year  \
2         Buffalo Bills          2.0            46.0      2013.0   
6         Buffalo Bills          3.0            73.0      2014.0   
13     Seattle Seahawks          2.0            63.0      2015.0   
37  Pittsburgh Steelers          3.0            82.0      2010.0   
53        New York Jets          4.0           126.0      2011.0   

   current_salary  hof_induction_year current_salary_nocommas  \
2       1,075,000                 NaN                 1075000   
6       1,762,000                 NaN                 1762000   
13        774,294                 NaN                  774294   
37      6,750,000                 NaN                 6750000   
53      3,750,000                 NaN                 3750000   

    current_salary_cleaned  
2                1075000.0  
6                1762000.0  
13                774294.0  
37               6750000.0  
53               3750000.0

## Creating dummy variables

We might, on occasion, want to "bin" or "discretize" a variable.  For example, we might want to take the previous dataframe and add dummy variables that map onto whether the salaries are "small" (< \\$1M) , "medium" (\\$1M - \\$10M), or "large" (> \\$10M). A **dummy variable** is a binary indicator variable (often used in regressions) to represent a categorical variable. 

We could do something like the following (I encourage you to make detailed notes about the following lines; we do several important steps.)

In [18]:
bins = [0,1000000,10000000,1000000000]

In [19]:
pd.cut(nfl_football_players_salaries['current_salary_cleaned'],bins,labels=['small','medium','large'])[0:10]

2      medium
6      medium
13      small
37     medium
53     medium
57      small
90      small
106    medium
113     small
114    medium
Name: current_salary_cleaned, dtype: category
Categories (3, object): ['small' < 'medium' < 'large']

In [20]:
dummies = pd.get_dummies(pd.cut(nfl_football_players_salaries['current_salary_cleaned'],bins,labels=['small','medium','large']))

In [21]:
dummies.head()

small  medium  large
2       0       1      0
6       0       1      0
13      1       0      0
37      0       1      0
53      0       1      0

pd.concat stacks together objects along an axis _(Section 8.2 McKinney)_

In [22]:
nfl_cats = pd.concat([nfl_football_players_salaries,dummies],axis=1)

In [23]:
nfl_cats.tail()

player_id             name position height  weight  \
24885      13753    Paxton Lynch        QB    6-7   244.0   
24917      22916    Jason Verrett       CB   5-10   189.0   
24923      18586  Jordan Richards       SS   5-11   211.0   
24967       7063   Michael Floyd        WR    6-3   220.0   
24984      10701      Chris Ivory       RB    6-0   222.0   

               current_team  birth_date      birth_place death_date  \
24885        Denver Broncos  1994-02-12  San Antonio, TX       None   
24917  Los Angeles Chargers  1991-05-13    Fairfield, CA       None   
24923  New England Patriots  1993-01-21   Sacramento, CA       None   
24967     Minnesota Vikings  1989-11-27     St. Paul, MN       None   
24984  Jacksonville Jaguars  1988-03-22     Longview, TX       None   

              college  ... draft_round draft_position  draft_year  \
24885         Memphis  ...         1.0           26.0      2016.0   
24917             TCU  ...         1.0           25.0      2014.0   
24923        Stanford  ...         2.0           64.0      2015.0   
24967      Notre Dame  ...         1.0           13.0      2012.0   
24984  Washington St.  ...         NaN            NaN         NaN   

       current_salary  hof_induction_year current_salary_nocommas  \
24885         880,741                 NaN                  880741   
24917         860,000                 NaN                  860000   
24923         772,413                 NaN                  772413   
24967         887,058                 NaN                  887058   
24984       4,500,000                 NaN                 4500000   

       current_salary_cleaned small  medium  large  
24885                880741.0     1       0      0  
24917                860000.0     1       0      0  
24923                772413.0     1       0      0  
24967                887058.0     1       0      0  
24984               4500000.0     0       1      0  

[5 rows x 22 columns]

## <font color="magenta">Q2 (12 points, 10 minutes): 
Create a new dataframe that contains all the columns in the nfl_football_players dataframe as well as an additional column that contains each player's height in centimeters. Show the first 5 rows of your result. </font>

**Hints:** 
- 1 inch = 2.54 cm
- you can use the vectorized string function str.split() to separate feet and inches from the original dataframe column
- remember to cast strings to numeric types if you're going to perform math on them
- you might want to create an intermediate (temporary) DataFrame to help you keep things clear instead of attempting to do 
this in one line

In [30]:
height_temp = nfl_football_players.height.str.split('-', expand=True)
height_temp.columns = ['feet', 'inch']
height_temp = height_temp.astype(float)
height_temp['height_cm'] = (height_temp['feet']*12 + height_temp['inch'])*2.54
nfl_cm = pd.concat([nfl_football_players,height_temp['height_cm']],axis=1)
nfl_cm.head(5)

player_id             name position height  weight    current_team  \
0       1809  Robert Blackmon       DB    6-0   208.0            None   
1      23586       Dean Wells       LB    6-3   248.0            None   
2        355      Kiko Alonso      ILB    6-3   238.0  Miami Dolphins   
3      18182     Steve Ramsey       QB    6-2   210.0            None   
4      16250       Cory Nelms       CB    6-0   195.0            None   

   birth_date     birth_place  death_date      college       high_school  \
0  1967-05-12    Bay City, TX        None       Baylor     Van Vleck, TX   
1  1970-07-20  Louisville, KY        None     Kentucky    Holy Cross, KY   
2  1990-08-14      Newton, MA        None       Oregon     Los Gatos, CA   
3  1948-04-22      Dallas, TX  1999-10-15  North Texas  W.W. Samuell, TX   
4  1988-02-27     Neptune, NJ        None   Miami (FL)       Neptune, NJ   

           draft_team  draft_round  draft_position  draft_year current_salary  \
0    Seattle Seahawks          2.0            34.0      1990.0           None   
1    Seattle Seahawks          4.0            85.0      1993.0           None   
2       Buffalo Bills          2.0            46.0      2013.0      1,075,000   
3  New Orleans Saints          5.0           126.0      1970.0           None   
4                None          NaN             NaN         NaN           None   

   hof_induction_year current_salary_nocommas  current_salary_cleaned  \
0                 NaN                    None                     NaN   
1                 NaN                    None                     NaN   
2                 NaN                 1075000               1075000.0   
3                 NaN                    None                     NaN   
4                 NaN                    None                     NaN   

   height_cm  
0     182.88  
1     190.50  
2     190.50  
3     187.96  
4     182.88

## Scraping Tables from HTML

The ```pd.read_html``` function returns a list of DataFrames read from an HTML source.  The following line will return a list of DataFrames from https://en.wikipedia.org/wiki/List_of_largest_sports_contracts

This part does not function in a stable way dues to the vpn issue, but fortunately it happened to succeed this time.

In [32]:
contracts_scraped = pd.read_html('https://en.wikipedia.org/wiki/List_of_largest_sports_contracts',header=0)  

In [33]:
len(contracts_scraped)

1

To get the first table, you'll need to pull off the 0th element:

In [34]:
contracts = contracts_scraped[0]
contracts.head()

Rank               Player Team/Streaming Service                 Sport  \
0    1     Lionel Messi[b]‡           FC Barcelona  Association football   
1    2  Patrick Mahomes[d]‡     Kansas City Chiefs     American football   
2    3           Mike Trout     Los Angeles Angels              Baseball   
3    4       Canelo Álvarez                  DAZN*                Boxing   
4    5         Mookie Betts    Los Angeles Dodgers              Baseball   

     Length of contract Contract value (USD) Average per year (USD)  \
0   4 years (2017–2021)         $673,919,105           $168,479,776   
1  12 years (2020-2031)         $503,000,000            $41,916,667   
2  12 years (2019–2030)         $426,500,000            $35,541,667   
3   5 years (2018–2023)         $365,000,000            $73,000,000   
4  12 years (2020–2032)         $365,000,000            $30,416,667   

  Average per fixture[a] (USD)  Ref  
0                $3,743,995[c]  [1]  
1                   $2,619,791  [2]  
2                     $219,393  [3]  
3                  $33,181,818  [4]  
4                     $187,757  [5]

## <font color="magenta"> Q3 (2 points): 
Count the number of players from each sport in the List of Largest Sports Contracts. 
Hint:  see value_counts() description above
<font>

In [35]:
contracts.Sport.value_counts()

Baseball                54
Basketball              36
American football        7
Association football     2
Boxing                   1
Auto racing              1
Name: Sport, dtype: int64

In [38]:
contracts.shape

(101, 9)

54 of the players come from baseball. As there are 101 players in total, this means more than half of the players are from baseball. 

# Joining, Combining, and Reshaping

In [36]:
frodo_url = 'https://en.wikipedia.org/wiki/Frodo_Baggins'
frodo_tables = pd.read_html(frodo_url)
frodo_tables[0]

Frodo Baggins                    Frodo Baggins.1
0         First appearance  The Fellowship of the Ring (1954)
1          Last appearance           Bilbo's Last Song (1974)
2  In-universe information            In-universe information
3                  Aliases                      Mr. Underhill
4                     Race                             Hobbit
5              Affiliation                Company of the Ring
6                   Family             Bilbo Baggins (cousin)
7                     Home                          The Shire

Now let's load the page for [Legolas](https://en.wikipedia.org/wiki/Legolas):

In [37]:
legolas_url = 'https://en.wikipedia.org/wiki/Legolas'
legolas_tables = pd.read_html(legolas_url)
legolas_tables[0]

Legolas  \
0  The Lord of the Rings character   
1          In-universe information   
2                          Aliases   
3                             Race   
4                           Gender   
5                      Affiliation   
6                             Home   
7                          Book(s)   

                                     Legolas.1  
0              The Lord of the Rings character  
1                      In-universe information  
2  Greenleaf, (Legolas translatedinto English)  
3                                   Sindar Elf  
4                                         Male  
5                          Company of the Ring  
6                                     Mirkwood  
7                        The Lord of the Rings

Hmmmm.  That doesn't look quite right.

Let's take a look at some URLs and figure out what's going on:
https://en.wikipedia.org/wiki/Legolas

### Inspect the Frodo and Legolas pages and see if you can figure out some _attributes_ of the table we're interested in.


Both Frodo page and Legolas page have a `class='infobox'` table, which means they have some shared characteristics that could be used to filter out the right tables we want.

You'll notice that there are some characteristics that the "Information" box share across pages.  We can leverage that 
information by using the ```attrs``` attribute of ```read_html```.  For example, if we wanted to extract  the element(s) that had
an ```id``` of ```info```, we could use

```pd.read_html(url,{'id':'info'})```



## <font color="magenta">Q4: (2 points) Fill in the following code block to extract only the "Information" table for the Legolas page:

In [40]:
a = {'class': 'infobox'} # create an appropriate dictionary
pd.read_html(legolas_url, attrs=a)

[                           Legolas  \
 0  The Lord of the Rings character   
 1          In-universe information   
 2                          Aliases   
 3                             Race   
 4                           Gender   
 5                      Affiliation   
 6                             Home   
 7                          Book(s)   
 
                                      Legolas.1  
 0              The Lord of the Rings character  
 1                      In-universe information  
 2  Greenleaf, (Legolas translatedinto English)  
 3                                   Sindar Elf  
 4                                         Male  
 5                          Company of the Ring  
 6                                     Mirkwood  
 7                        The Lord of the Rings  ]

Now let's define a function that, given a Wikipedia URL, will extract the contents of the Aliases component of the infobox table:

In [41]:
def get_aliases(url):
    try:
        tables = pd.read_html(url, attrs={'class':'infobox'}) # extract only tables with class=infobox
        print(url,len(tables))   # sanity check: we should have just 1 table
        infotable = tables[0]    # pull the first table into a DataFrame
        ret = ''                 # initialize an empty string for our return value
        x = infotable.set_index(infotable.columns[0]).loc['Aliases'] # setting the index on column 0 will allow us to use .loc to look up the value of 'Aliases'
        ret = x.values[0]
    except:
        print("Problem with " + url)
        ret = 'None'
    return ret

And let's try it out:

In [42]:
get_aliases(legolas_url)

https://en.wikipedia.org/wiki/Legolas 1


'Greenleaf, (Legolas translatedinto English)'

So far, so good.  It seems to work.  Now let's set up a DataFrame with a bunch of URLs:

In [43]:
urls = ['https://en.wikipedia.org/wiki/Gimli_(Middle-earth)',
        'https://en.wikipedia.org/wiki/Frodo_Baggins',
        'https://en.wikipedia.org/wiki/Legolas',
        'https://en.wikipedia.org/wiki/Peregrin_Took',
        'https://en.wikipedia.org/wiki/Meriadoc_Brandybuck']
names = ['Gimli',
         'Frodo',
         'Legolas',
         'Pippin',
         'Meriadoc']

In [44]:
url_df = pd.DataFrame()
url_df['name'] = names
url_df['url'] = urls

In [45]:
url_df

name                                                url
0     Gimli  https://en.wikipedia.org/wiki/Gimli_(Middle-ea...
1     Frodo        https://en.wikipedia.org/wiki/Frodo_Baggins
2   Legolas              https://en.wikipedia.org/wiki/Legolas
3    Pippin        https://en.wikipedia.org/wiki/Peregrin_Took
4  Meriadoc  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck

The pythonic way of iterating through each of those rows would involve the use of some sort of ```for``` loop.  In pandas,
however, you can use the ```apply``` function to process an entire column!

In [46]:
url_df['url'].apply(get_aliases)

https://en.wikipedia.org/wiki/Gimli_(Middle-earth) 1
https://en.wikipedia.org/wiki/Frodo_Baggins 1
https://en.wikipedia.org/wiki/Legolas 1
https://en.wikipedia.org/wiki/Peregrin_Took 1
https://en.wikipedia.org/wiki/Meriadoc_Brandybuck 1


0    Elf-friend Lockbearer Lord of the Glittering C...
1                                        Mr. Underhill
2          Greenleaf, (Legolas translatedinto English)
3                                               Pippin
4                             Merry,Master of Buckland
Name: url, dtype: object

We can take the resulting Series and assign it to a new column in our DataFrame:

In [47]:
url_df['aliases'] = url_df['url'].apply(get_aliases)

https://en.wikipedia.org/wiki/Gimli_(Middle-earth) 1
https://en.wikipedia.org/wiki/Frodo_Baggins 1
https://en.wikipedia.org/wiki/Legolas 1
https://en.wikipedia.org/wiki/Peregrin_Took 1
Problem with https://en.wikipedia.org/wiki/Meriadoc_Brandybuck


In [48]:
url_df

name                                                url  \
0     Gimli  https://en.wikipedia.org/wiki/Gimli_(Middle-ea...   
1     Frodo        https://en.wikipedia.org/wiki/Frodo_Baggins   
2   Legolas              https://en.wikipedia.org/wiki/Legolas   
3    Pippin        https://en.wikipedia.org/wiki/Peregrin_Took   
4  Meriadoc  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck   

                                             aliases  
0  Elf-friend Lockbearer Lord of the Glittering C...  
1                                      Mr. Underhill  
2        Greenleaf, (Legolas translatedinto English)  
3                                             Pippin  
4                                               None

Let's just put the ```url_df``` DataFrame aside for now.  We'll return to it later.

## Creating DataFrames and Exploring Indexes

Let's create some lists of data that we can use to construct a DataFrame:

In [49]:
names = ['Gandalf',
         'Gimli',
         'Frodo',
         'Legolas',
         'Bilbo',
         'Sam',
         'Pippin',
         'Boromir',
         'Aragorn',
         'Galadriel',
         'Meriadoc',
        'Lily']
races = ['Maia',
         'Dwarf',
         'Hobbit',
         'Elf',
         'Hobbit',
         'Hobbit',
         'Hobbit',
         'Man',
         'Man',
         'Elf',
         'Hobbit',
        'Hobbit']
magic = [10, 1, 4, 6, 4, 2, 0, 0, 2, 9, 0, np.NaN]
aggression = [7, 10, 2, 5, 1, 6, 3, 8, 7, 2, 4, np.NaN ]
stealth = [8, 2, 5, 10, 5, 4 ,5, 3, 9, 10, 6, np.NaN]

There are a few different ways to construct a DataFrame.  We can either use an empty constructor and assign Series:

## <font color="magenta"> Q5: (2 points) Construct a dataframe with 5 columns (names, races, magic, aggression, and stealth) using the lists above.

In [50]:
df = pd.DataFrame()
df['name'] = names
df['race'] = races
df['magic'] = magic
df['aggression'] = aggression
df['stealth'] = stealth

In [51]:
df

name    race  magic  aggression  stealth
0     Gandalf    Maia   10.0         7.0      8.0
1       Gimli   Dwarf    1.0        10.0      2.0
2       Frodo  Hobbit    4.0         2.0      5.0
3     Legolas     Elf    6.0         5.0     10.0
4       Bilbo  Hobbit    4.0         1.0      5.0
5         Sam  Hobbit    2.0         6.0      4.0
6      Pippin  Hobbit    0.0         3.0      5.0
7     Boromir     Man    0.0         8.0      3.0
8     Aragorn     Man    2.0         7.0      9.0
9   Galadriel     Elf    9.0         2.0     10.0
10   Meriadoc  Hobbit    0.0         4.0      6.0
11       Lily  Hobbit    NaN         NaN      NaN

Alternatively, we could have set things up with a dict:

In [52]:
df = pd.DataFrame({'name': names,'race':races,'magic':magic,'aggression': aggression,'stealth':stealth})

In [53]:
df.head()

name    race  magic  aggression  stealth
0  Gandalf    Maia   10.0         7.0      8.0
1    Gimli   Dwarf    1.0        10.0      2.0
2    Frodo  Hobbit    4.0         2.0      5.0
3  Legolas     Elf    6.0         5.0     10.0
4    Bilbo  Hobbit    4.0         1.0      5.0

Let's take a look at the index on the resulting DataFrame:

In [54]:
df.index

RangeIndex(start=0, stop=12, step=1)

We can set the index to something more useful than the default RangeIndex:

In [55]:
df_nameindexed = df.set_index('name')

And if we take a look at the results, we see that we have a pandas Index instead of a RangeIndex:

In [56]:
df_nameindexed.index

Index(['Gandalf', 'Gimli', 'Frodo', 'Legolas', 'Bilbo', 'Sam', 'Pippin',
       'Boromir', 'Aragorn', 'Galadriel', 'Meriadoc', 'Lily'],
      dtype='object', name='name')

In [57]:
df_nameindexed

race  magic  aggression  stealth
name                                         
Gandalf      Maia   10.0         7.0      8.0
Gimli       Dwarf    1.0        10.0      2.0
Frodo      Hobbit    4.0         2.0      5.0
Legolas       Elf    6.0         5.0     10.0
Bilbo      Hobbit    4.0         1.0      5.0
Sam        Hobbit    2.0         6.0      4.0
Pippin     Hobbit    0.0         3.0      5.0
Boromir       Man    0.0         8.0      3.0
Aragorn       Man    2.0         7.0      9.0
Galadriel     Elf    9.0         2.0     10.0
Meriadoc   Hobbit    0.0         4.0      6.0
Lily       Hobbit    NaN         NaN      NaN

Setting the name Series as the index allows us to do things like:

In [58]:
df_nameindexed.loc['Aragorn']

race          Man
magic         2.0
aggression    7.0
stealth       9.0
Name: Aragorn, dtype: object

Now recall the Hierarchical indexing from the readings. _(Section 8.1 McKinney)_ We can pass a list of column names to set_index to create a Hierarchical Index:

In [59]:
df_racename_indexed = df.set_index(['race','name'])

In [60]:
df_racename_indexed.index

MultiIndex([(  'Maia',   'Gandalf'),
            ( 'Dwarf',     'Gimli'),
            ('Hobbit',     'Frodo'),
            (   'Elf',   'Legolas'),
            ('Hobbit',     'Bilbo'),
            ('Hobbit',       'Sam'),
            ('Hobbit',    'Pippin'),
            (   'Man',   'Boromir'),
            (   'Man',   'Aragorn'),
            (   'Elf', 'Galadriel'),
            ('Hobbit',  'Meriadoc'),
            ('Hobbit',      'Lily')],
           names=['race', 'name'])

This will allow us to get a DataFrame that matches a value on the outer index:

In [61]:
df_racename_indexed.loc['Hobbit']

magic  aggression  stealth
name                                
Frodo       4.0         2.0      5.0
Bilbo       4.0         1.0      5.0
Sam         2.0         6.0      4.0
Pippin      0.0         3.0      5.0
Meriadoc    0.0         4.0      6.0
Lily        NaN         NaN      NaN

We can also use the index on a Series to match the outer index:

In [62]:
df_racename_indexed['magic'].loc['Hobbit']

name
Frodo       4.0
Bilbo       4.0
Sam         2.0
Pippin      0.0
Meriadoc    0.0
Lily        NaN
Name: magic, dtype: float64

Or both indexes:

In [63]:
df_racename_indexed['magic'].loc['Hobbit','Frodo']

4.0

Or just the inner index:

In [64]:
df_racename_indexed['magic'].loc[:,'Frodo']

race
Hobbit    4.0
Name: magic, dtype: float64

## <font color="magenta"> Q6: (2 points) Using .loc find how much aggression Legolas, an Elf, has.

In [66]:
df_racename_indexed['aggression'].loc['Elf', 'Legolas']

5.0

## Stacking and Unstacking

Stacking takes "wide" data and makes it "taller"

In [67]:
df.set_index(['race']).stack()

race              
Maia    name            Gandalf
        magic              10.0
        aggression          7.0
        stealth             8.0
Dwarf   name              Gimli
        magic               1.0
        aggression         10.0
        stealth             2.0
Hobbit  name              Frodo
        magic               4.0
        aggression          2.0
        stealth             5.0
Elf     name            Legolas
        magic               6.0
        aggression          5.0
        stealth            10.0
Hobbit  name              Bilbo
        magic               4.0
        aggression          1.0
        stealth             5.0
        name                Sam
        magic               2.0
        aggression          6.0
        stealth             4.0
        name             Pippin
        magic               0.0
        aggression          3.0
        stealth             5.0
Man     name            Boromir
        magic               0.0
        aggression   

If we call reset_index on the resulting Series, we get the following DataFrame:

In [68]:
df.set_index(['race']).stack().reset_index()

race     level_1          0
0     Maia        name    Gandalf
1     Maia       magic       10.0
2     Maia  aggression        7.0
3     Maia     stealth        8.0
4    Dwarf        name      Gimli
5    Dwarf       magic        1.0
6    Dwarf  aggression       10.0
7    Dwarf     stealth        2.0
8   Hobbit        name      Frodo
9   Hobbit       magic        4.0
10  Hobbit  aggression        2.0
11  Hobbit     stealth        5.0
12     Elf        name    Legolas
13     Elf       magic        6.0
14     Elf  aggression        5.0
15     Elf     stealth       10.0
16  Hobbit        name      Bilbo
17  Hobbit       magic        4.0
18  Hobbit  aggression        1.0
19  Hobbit     stealth        5.0
20  Hobbit        name        Sam
21  Hobbit       magic        2.0
22  Hobbit  aggression        6.0
23  Hobbit     stealth        4.0
24  Hobbit        name     Pippin
25  Hobbit       magic        0.0
26  Hobbit  aggression        3.0
27  Hobbit     stealth        5.0
28     Man        name    Boromir
29     Man       magic        0.0
30     Man  aggression        8.0
31     Man     stealth        3.0
32     Man        name    Aragorn
33     Man       magic        2.0
34     Man  aggression        7.0
35     Man     stealth        9.0
36     Elf        name  Galadriel
37     Elf       magic        9.0
38     Elf  aggression        2.0
39     Elf     stealth       10.0
40  Hobbit        name   Meriadoc
41  Hobbit       magic        0.0
42  Hobbit  aggression        4.0
43  Hobbit     stealth        6.0
44  Hobbit        name       Lily

The column names in the above DataFrame aren't particularly helpful, so we can rename them:

In [69]:
df.set_index(['race']).stack().reset_index().rename(columns = {'level_0':'ID','level_1':'variable',0:'value'})

race    variable      value
0     Maia        name    Gandalf
1     Maia       magic       10.0
2     Maia  aggression        7.0
3     Maia     stealth        8.0
4    Dwarf        name      Gimli
5    Dwarf       magic        1.0
6    Dwarf  aggression       10.0
7    Dwarf     stealth        2.0
8   Hobbit        name      Frodo
9   Hobbit       magic        4.0
10  Hobbit  aggression        2.0
11  Hobbit     stealth        5.0
12     Elf        name    Legolas
13     Elf       magic        6.0
14     Elf  aggression        5.0
15     Elf     stealth       10.0
16  Hobbit        name      Bilbo
17  Hobbit       magic        4.0
18  Hobbit  aggression        1.0
19  Hobbit     stealth        5.0
20  Hobbit        name        Sam
21  Hobbit       magic        2.0
22  Hobbit  aggression        6.0
23  Hobbit     stealth        4.0
24  Hobbit        name     Pippin
25  Hobbit       magic        0.0
26  Hobbit  aggression        3.0
27  Hobbit     stealth        5.0
28     Man        name    Boromir
29     Man       magic        0.0
30     Man  aggression        8.0
31     Man     stealth        3.0
32     Man        name    Aragorn
33     Man       magic        2.0
34     Man  aggression        7.0
35     Man     stealth        9.0
36     Elf        name  Galadriel
37     Elf       magic        9.0
38     Elf  aggression        2.0
39     Elf     stealth       10.0
40  Hobbit        name   Meriadoc
41  Hobbit       magic        0.0
42  Hobbit  aggression        4.0
43  Hobbit     stealth        6.0
44  Hobbit        name       Lily

You can do the opposite of stacking by using the ```unstack``` function:

In [70]:
df_stacked = df.stack()

In [71]:
df_stacked

0   name            Gandalf
    race               Maia
    magic              10.0
    aggression          7.0
    stealth             8.0
1   name              Gimli
    race              Dwarf
    magic               1.0
    aggression         10.0
    stealth             2.0
2   name              Frodo
    race             Hobbit
    magic               4.0
    aggression          2.0
    stealth             5.0
3   name            Legolas
    race                Elf
    magic               6.0
    aggression          5.0
    stealth            10.0
4   name              Bilbo
    race             Hobbit
    magic               4.0
    aggression          1.0
    stealth             5.0
5   name                Sam
    race             Hobbit
    magic               2.0
    aggression          6.0
    stealth             4.0
6   name             Pippin
    race             Hobbit
    magic               0.0
    aggression          3.0
    stealth             5.0
7   name            

In [72]:
df_stacked.unstack()

name    race magic aggression stealth
0     Gandalf    Maia  10.0        7.0     8.0
1       Gimli   Dwarf   1.0       10.0     2.0
2       Frodo  Hobbit   4.0        2.0     5.0
3     Legolas     Elf   6.0        5.0    10.0
4       Bilbo  Hobbit   4.0        1.0     5.0
5         Sam  Hobbit   2.0        6.0     4.0
6      Pippin  Hobbit   0.0        3.0     5.0
7     Boromir     Man   0.0        8.0     3.0
8     Aragorn     Man   2.0        7.0     9.0
9   Galadriel     Elf   9.0        2.0    10.0
10   Meriadoc  Hobbit   0.0        4.0     6.0
11       Lily  Hobbit   NaN        NaN     NaN

Why would we want to stack or unstack?  It depends on what sorts of analyses we want to do "downstream".  It's also the basis for pivoting, melting, and pivot tables, which we'll cover later.

## Joining Data



Let's say we have another CSV file that contains URLs to Wikipedia pages for some of the LOTR characters:

In [73]:
urls = pd.read_csv('https://github.com/umsi-data-science/data/raw/main/lotr_wikipedia.csv')
urls

name                                                url
0     Gandalf              https://en.wikipedia.org/wiki/Gandalf
1       Gimli  https://en.wikipedia.org/wiki/Gimli_(Middle-ea...
2       Frodo        https://en.wikipedia.org/wiki/Frodo_Baggins
3     Legolas              https://en.wikipedia.org/wiki/Legolas
4       Bilbo        https://en.wikipedia.org/wiki/Bilbo_Baggins
5         Sam       https://en.wikipedia.org/wiki/Samwise_Gamgee
6      Pippin        https://en.wikipedia.org/wiki/Peregrin_Took
7     Boromir              https://en.wikipedia.org/wiki/Boromir
8     Aragorn              https://en.wikipedia.org/wiki/Aragorn
9   Galadriel            https://en.wikipedia.org/wiki/Galadriel
10   Meriadoc  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck
11       Lily                                                NaN

It looks like the rows are "aligned", so we can use the ```concat``` function to concatenate the two DataFrames.
Note that we specify the axis to be the columns.  The default is to concatenate by rows, which isn't what we want.

In [74]:
pd.concat([df,urls],axis="columns")

name    race  magic  aggression  stealth       name  \
0     Gandalf    Maia   10.0         7.0      8.0    Gandalf   
1       Gimli   Dwarf    1.0        10.0      2.0      Gimli   
2       Frodo  Hobbit    4.0         2.0      5.0      Frodo   
3     Legolas     Elf    6.0         5.0     10.0    Legolas   
4       Bilbo  Hobbit    4.0         1.0      5.0      Bilbo   
5         Sam  Hobbit    2.0         6.0      4.0        Sam   
6      Pippin  Hobbit    0.0         3.0      5.0     Pippin   
7     Boromir     Man    0.0         8.0      3.0    Boromir   
8     Aragorn     Man    2.0         7.0      9.0    Aragorn   
9   Galadriel     Elf    9.0         2.0     10.0  Galadriel   
10   Meriadoc  Hobbit    0.0         4.0      6.0   Meriadoc   
11       Lily  Hobbit    NaN         NaN      NaN       Lily   

                                                  url  
0               https://en.wikipedia.org/wiki/Gandalf  
1   https://en.wikipedia.org/wiki/Gimli_(Middle-ea...  
2         https://en.wikipedia.org/wiki/Frodo_Baggins  
3               https://en.wikipedia.org/wiki/Legolas  
4         https://en.wikipedia.org/wiki/Bilbo_Baggins  
5        https://en.wikipedia.org/wiki/Samwise_Gamgee  
6         https://en.wikipedia.org/wiki/Peregrin_Took  
7               https://en.wikipedia.org/wiki/Boromir  
8               https://en.wikipedia.org/wiki/Aragorn  
9             https://en.wikipedia.org/wiki/Galadriel  
10  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck  
11                                                NaN

That's great, and it's consistent with what we've used in previous classes.  But what happens if the 
rows in the two DataFrames don't match up?  Let's load another file that has a slightly different
sequence of rows:

## <font color="magenta"> Q7: (2 points) Construct a dataframe called ```urls_wrong_order``` with https://github.com/umsi-data-science/data/raw/main/lotr_wikipedia_wrong_order.csv and concat it with df.

In [76]:
urls_wrong_order = pd.read_csv('https://github.com/umsi-data-science/data/raw/main/lotr_wikipedia_wrong_order.csv')
pd.concat([df,urls_wrong_order],axis="columns")

name    race  magic  aggression  stealth       name  \
0     Gandalf    Maia   10.0         7.0      8.0    Boromir   
1       Gimli   Dwarf    1.0        10.0      2.0    Aragorn   
2       Frodo  Hobbit    4.0         2.0      5.0    Gandalf   
3     Legolas     Elf    6.0         5.0     10.0   Meriadoc   
4       Bilbo  Hobbit    4.0         1.0      5.0      Gimli   
5         Sam  Hobbit    2.0         6.0      4.0    Legolas   
6      Pippin  Hobbit    0.0         3.0      5.0      Bilbo   
7     Boromir     Man    0.0         8.0      3.0     Pippin   
8     Aragorn     Man    2.0         7.0      9.0  Galadriel   
9   Galadriel     Elf    9.0         2.0     10.0       Lily   
10   Meriadoc  Hobbit    0.0         4.0      6.0      Frodo   
11       Lily  Hobbit    NaN         NaN      NaN        Sam   

                                                  url  
0               https://en.wikipedia.org/wiki/Boromir  
1               https://en.wikipedia.org/wiki/Aragorn  
2               https://en.wikipedia.org/wiki/Gandalf  
3   https://en.wikipedia.org/wiki/Meriadoc_Brandybuck  
4   https://en.wikipedia.org/wiki/Gimli_(Middle-ea...  
5               https://en.wikipedia.org/wiki/Legolas  
6         https://en.wikipedia.org/wiki/Bilbo_Baggins  
7         https://en.wikipedia.org/wiki/Peregrin_Took  
8             https://en.wikipedia.org/wiki/Galadriel  
9                                                 NaN  
10        https://en.wikipedia.org/wiki/Frodo_Baggins  
11       https://en.wikipedia.org/wiki/Samwise_Gamgee

Take a closer look at the name and url columns.  Something's not quite right.

We can work around that by using the appropriate indexing and then using the SQL-like ```merge``` function.

In [77]:
df_names = df.set_index('name')

In [78]:
df_names

race  magic  aggression  stealth
name                                         
Gandalf      Maia   10.0         7.0      8.0
Gimli       Dwarf    1.0        10.0      2.0
Frodo      Hobbit    4.0         2.0      5.0
Legolas       Elf    6.0         5.0     10.0
Bilbo      Hobbit    4.0         1.0      5.0
Sam        Hobbit    2.0         6.0      4.0
Pippin     Hobbit    0.0         3.0      5.0
Boromir       Man    0.0         8.0      3.0
Aragorn       Man    2.0         7.0      9.0
Galadriel     Elf    9.0         2.0     10.0
Meriadoc   Hobbit    0.0         4.0      6.0
Lily       Hobbit    NaN         NaN      NaN

In [79]:
urls_wrong_order_names = urls_wrong_order.set_index('name')
urls_wrong_order_names

url
name                                                        
Boromir                https://en.wikipedia.org/wiki/Boromir
Aragorn                https://en.wikipedia.org/wiki/Aragorn
Gandalf                https://en.wikipedia.org/wiki/Gandalf
Meriadoc   https://en.wikipedia.org/wiki/Meriadoc_Brandybuck
Gimli      https://en.wikipedia.org/wiki/Gimli_(Middle-ea...
Legolas                https://en.wikipedia.org/wiki/Legolas
Bilbo            https://en.wikipedia.org/wiki/Bilbo_Baggins
Pippin           https://en.wikipedia.org/wiki/Peregrin_Took
Galadriel            https://en.wikipedia.org/wiki/Galadriel
Lily                                                     NaN
Frodo            https://en.wikipedia.org/wiki/Frodo_Baggins
Sam             https://en.wikipedia.org/wiki/Samwise_Gamgee

In [80]:
df_names.join(urls_wrong_order_names)

race  magic  aggression  stealth  \
name                                            
Gandalf      Maia   10.0         7.0      8.0   
Gimli       Dwarf    1.0        10.0      2.0   
Frodo      Hobbit    4.0         2.0      5.0   
Legolas       Elf    6.0         5.0     10.0   
Bilbo      Hobbit    4.0         1.0      5.0   
Sam        Hobbit    2.0         6.0      4.0   
Pippin     Hobbit    0.0         3.0      5.0   
Boromir       Man    0.0         8.0      3.0   
Aragorn       Man    2.0         7.0      9.0   
Galadriel     Elf    9.0         2.0     10.0   
Meriadoc   Hobbit    0.0         4.0      6.0   
Lily       Hobbit    NaN         NaN      NaN   

                                                         url  
name                                                          
Gandalf                https://en.wikipedia.org/wiki/Gandalf  
Gimli      https://en.wikipedia.org/wiki/Gimli_(Middle-ea...  
Frodo            https://en.wikipedia.org/wiki/Frodo_Baggins  
Legolas                https://en.wikipedia.org/wiki/Legolas  
Bilbo            https://en.wikipedia.org/wiki/Bilbo_Baggins  
Sam             https://en.wikipedia.org/wiki/Samwise_Gamgee  
Pippin           https://en.wikipedia.org/wiki/Peregrin_Took  
Boromir                https://en.wikipedia.org/wiki/Boromir  
Aragorn                https://en.wikipedia.org/wiki/Aragorn  
Galadriel            https://en.wikipedia.org/wiki/Galadriel  
Meriadoc   https://en.wikipedia.org/wiki/Meriadoc_Brandybuck  
Lily                                                     NaN

In [81]:
df.head()

name    race  magic  aggression  stealth
0  Gandalf    Maia   10.0         7.0      8.0
1    Gimli   Dwarf    1.0        10.0      2.0
2    Frodo  Hobbit    4.0         2.0      5.0
3  Legolas     Elf    6.0         5.0     10.0
4    Bilbo  Hobbit    4.0         1.0      5.0

In [82]:
urls_wrong_order.head()

name                                                url
0   Boromir              https://en.wikipedia.org/wiki/Boromir
1   Aragorn              https://en.wikipedia.org/wiki/Aragorn
2   Gandalf              https://en.wikipedia.org/wiki/Gandalf
3  Meriadoc  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck
4     Gimli  https://en.wikipedia.org/wiki/Gimli_(Middle-ea...

In [83]:
urls_wrong_order['name']

0       Boromir
1       Aragorn
2       Gandalf
3      Meriadoc
4         Gimli
5       Legolas
6         Bilbo
7        Pippin
8     Galadriel
9          Lily
10        Frodo
11          Sam
Name: name, dtype: object

In [84]:
df['name']

0       Gandalf
1         Gimli
2         Frodo
3       Legolas
4         Bilbo
5           Sam
6        Pippin
7       Boromir
8       Aragorn
9     Galadriel
10     Meriadoc
11         Lily
Name: name, dtype: object

In [85]:
df.merge(urls_wrong_order,on='name')

name    race  magic  aggression  stealth  \
0     Gandalf    Maia   10.0         7.0      8.0   
1       Gimli   Dwarf    1.0        10.0      2.0   
2       Frodo  Hobbit    4.0         2.0      5.0   
3     Legolas     Elf    6.0         5.0     10.0   
4       Bilbo  Hobbit    4.0         1.0      5.0   
5         Sam  Hobbit    2.0         6.0      4.0   
6      Pippin  Hobbit    0.0         3.0      5.0   
7     Boromir     Man    0.0         8.0      3.0   
8     Aragorn     Man    2.0         7.0      9.0   
9   Galadriel     Elf    9.0         2.0     10.0   
10   Meriadoc  Hobbit    0.0         4.0      6.0   
11       Lily  Hobbit    NaN         NaN      NaN   

                                                  url  
0               https://en.wikipedia.org/wiki/Gandalf  
1   https://en.wikipedia.org/wiki/Gimli_(Middle-ea...  
2         https://en.wikipedia.org/wiki/Frodo_Baggins  
3               https://en.wikipedia.org/wiki/Legolas  
4         https://en.wikipedia.org/wiki/Bilbo_Baggins  
5        https://en.wikipedia.org/wiki/Samwise_Gamgee  
6         https://en.wikipedia.org/wiki/Peregrin_Took  
7               https://en.wikipedia.org/wiki/Boromir  
8               https://en.wikipedia.org/wiki/Aragorn  
9             https://en.wikipedia.org/wiki/Galadriel  
10  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck  
11                                                NaN

Now let's add a few additional URLs:

In [86]:
urls_extras = pd.read_csv("https://github.com/umsi-data-science/data/raw/main/lotr_wikipedia_extras.csv")

In [87]:
urls_extras

name                                      url
0  Treebeard  https://en.wikipedia.org/wiki/Treebeard
1     Elrond     https://en.wikipedia.org/wiki/Elrond

And now let's use concat to add the new entries to the DataFrame.

In [88]:
urls_complete = pd.concat([urls,urls_extras])

In [89]:
urls_complete

name                                                url
0     Gandalf              https://en.wikipedia.org/wiki/Gandalf
1       Gimli  https://en.wikipedia.org/wiki/Gimli_(Middle-ea...
2       Frodo        https://en.wikipedia.org/wiki/Frodo_Baggins
3     Legolas              https://en.wikipedia.org/wiki/Legolas
4       Bilbo        https://en.wikipedia.org/wiki/Bilbo_Baggins
5         Sam       https://en.wikipedia.org/wiki/Samwise_Gamgee
6      Pippin        https://en.wikipedia.org/wiki/Peregrin_Took
7     Boromir              https://en.wikipedia.org/wiki/Boromir
8     Aragorn              https://en.wikipedia.org/wiki/Aragorn
9   Galadriel            https://en.wikipedia.org/wiki/Galadriel
10   Meriadoc  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck
11       Lily                                                NaN
0   Treebeard            https://en.wikipedia.org/wiki/Treebeard
1      Elrond               https://en.wikipedia.org/wiki/Elrond

Now that we've got a complete (for our purposes) list of URLs, let's use that DataFrame and our original
one to demonstrate the different types of ```join```s.

By default, ```join``` uses a left join, which means the all the values from the "left"
side are used, whether or not there's a corresponding entry from the "right" side.  In the example 
below, note that the url value for "Lily" is "NaN":

In [90]:
df.merge(urls_wrong_order,on='name',how='left')

name    race  magic  aggression  stealth  \
0     Gandalf    Maia   10.0         7.0      8.0   
1       Gimli   Dwarf    1.0        10.0      2.0   
2       Frodo  Hobbit    4.0         2.0      5.0   
3     Legolas     Elf    6.0         5.0     10.0   
4       Bilbo  Hobbit    4.0         1.0      5.0   
5         Sam  Hobbit    2.0         6.0      4.0   
6      Pippin  Hobbit    0.0         3.0      5.0   
7     Boromir     Man    0.0         8.0      3.0   
8     Aragorn     Man    2.0         7.0      9.0   
9   Galadriel     Elf    9.0         2.0     10.0   
10   Meriadoc  Hobbit    0.0         4.0      6.0   
11       Lily  Hobbit    NaN         NaN      NaN   

                                                  url  
0               https://en.wikipedia.org/wiki/Gandalf  
1   https://en.wikipedia.org/wiki/Gimli_(Middle-ea...  
2         https://en.wikipedia.org/wiki/Frodo_Baggins  
3               https://en.wikipedia.org/wiki/Legolas  
4         https://en.wikipedia.org/wiki/Bilbo_Baggins  
5        https://en.wikipedia.org/wiki/Samwise_Gamgee  
6         https://en.wikipedia.org/wiki/Peregrin_Took  
7               https://en.wikipedia.org/wiki/Boromir  
8               https://en.wikipedia.org/wiki/Aragorn  
9             https://en.wikipedia.org/wiki/Galadriel  
10  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck  
11                                                NaN

The "opposite" of a left join is, perhaps unsurprisingly, a "right" join, in which
all the values from the "right" side are used, whether or not a corresponding
value from the "left" side exists. Note in the following example that "Lily" has
disappeared, and Treebeard and Elrond lack information about "race", "magic", "aggression", and "stealth".

In [91]:
df.merge(urls_complete,on='name',how='right')

name    race  magic  aggression  stealth  \
0     Gandalf    Maia   10.0         7.0      8.0   
1       Gimli   Dwarf    1.0        10.0      2.0   
2       Frodo  Hobbit    4.0         2.0      5.0   
3     Legolas     Elf    6.0         5.0     10.0   
4       Bilbo  Hobbit    4.0         1.0      5.0   
5         Sam  Hobbit    2.0         6.0      4.0   
6      Pippin  Hobbit    0.0         3.0      5.0   
7     Boromir     Man    0.0         8.0      3.0   
8     Aragorn     Man    2.0         7.0      9.0   
9   Galadriel     Elf    9.0         2.0     10.0   
10   Meriadoc  Hobbit    0.0         4.0      6.0   
11       Lily  Hobbit    NaN         NaN      NaN   
12  Treebeard     NaN    NaN         NaN      NaN   
13     Elrond     NaN    NaN         NaN      NaN   

                                                  url  
0               https://en.wikipedia.org/wiki/Gandalf  
1   https://en.wikipedia.org/wiki/Gimli_(Middle-ea...  
2         https://en.wikipedia.org/wiki/Frodo_Baggins  
3               https://en.wikipedia.org/wiki/Legolas  
4         https://en.wikipedia.org/wiki/Bilbo_Baggins  
5        https://en.wikipedia.org/wiki/Samwise_Gamgee  
6         https://en.wikipedia.org/wiki/Peregrin_Took  
7               https://en.wikipedia.org/wiki/Boromir  
8               https://en.wikipedia.org/wiki/Aragorn  
9             https://en.wikipedia.org/wiki/Galadriel  
10  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck  
11                                                NaN  
12            https://en.wikipedia.org/wiki/Treebeard  
13               https://en.wikipedia.org/wiki/Elrond

In addition to "left" and "right" joins, we have "outer" joins, which include
values from both the "left" and "right" DataFrames, regardless of whether
there are corresponding values in the other DataFrame.  Note that all of 
"Lily", "Treebeard" and "Elrond" are present in the following DataFrame:

In [92]:
df.merge(urls_complete,on='name',how='outer')

name    race  magic  aggression  stealth  \
0     Gandalf    Maia   10.0         7.0      8.0   
1       Gimli   Dwarf    1.0        10.0      2.0   
2       Frodo  Hobbit    4.0         2.0      5.0   
3     Legolas     Elf    6.0         5.0     10.0   
4       Bilbo  Hobbit    4.0         1.0      5.0   
5         Sam  Hobbit    2.0         6.0      4.0   
6      Pippin  Hobbit    0.0         3.0      5.0   
7     Boromir     Man    0.0         8.0      3.0   
8     Aragorn     Man    2.0         7.0      9.0   
9   Galadriel     Elf    9.0         2.0     10.0   
10   Meriadoc  Hobbit    0.0         4.0      6.0   
11       Lily  Hobbit    NaN         NaN      NaN   
12  Treebeard     NaN    NaN         NaN      NaN   
13     Elrond     NaN    NaN         NaN      NaN   

                                                  url  
0               https://en.wikipedia.org/wiki/Gandalf  
1   https://en.wikipedia.org/wiki/Gimli_(Middle-ea...  
2         https://en.wikipedia.org/wiki/Frodo_Baggins  
3               https://en.wikipedia.org/wiki/Legolas  
4         https://en.wikipedia.org/wiki/Bilbo_Baggins  
5        https://en.wikipedia.org/wiki/Samwise_Gamgee  
6         https://en.wikipedia.org/wiki/Peregrin_Took  
7               https://en.wikipedia.org/wiki/Boromir  
8               https://en.wikipedia.org/wiki/Aragorn  
9             https://en.wikipedia.org/wiki/Galadriel  
10  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck  
11                                                NaN  
12            https://en.wikipedia.org/wiki/Treebeard  
13               https://en.wikipedia.org/wiki/Elrond

Finally, there are "inner" joins, which include only those values that exist in both the "left" and "right" DataFrames:

In [93]:
df.merge(urls_complete,on='name',how='inner')

name    race  magic  aggression  stealth  \
0     Gandalf    Maia   10.0         7.0      8.0   
1       Gimli   Dwarf    1.0        10.0      2.0   
2       Frodo  Hobbit    4.0         2.0      5.0   
3     Legolas     Elf    6.0         5.0     10.0   
4       Bilbo  Hobbit    4.0         1.0      5.0   
5         Sam  Hobbit    2.0         6.0      4.0   
6      Pippin  Hobbit    0.0         3.0      5.0   
7     Boromir     Man    0.0         8.0      3.0   
8     Aragorn     Man    2.0         7.0      9.0   
9   Galadriel     Elf    9.0         2.0     10.0   
10   Meriadoc  Hobbit    0.0         4.0      6.0   
11       Lily  Hobbit    NaN         NaN      NaN   

                                                  url  
0               https://en.wikipedia.org/wiki/Gandalf  
1   https://en.wikipedia.org/wiki/Gimli_(Middle-ea...  
2         https://en.wikipedia.org/wiki/Frodo_Baggins  
3               https://en.wikipedia.org/wiki/Legolas  
4         https://en.wikipedia.org/wiki/Bilbo_Baggins  
5        https://en.wikipedia.org/wiki/Samwise_Gamgee  
6         https://en.wikipedia.org/wiki/Peregrin_Took  
7               https://en.wikipedia.org/wiki/Boromir  
8               https://en.wikipedia.org/wiki/Aragorn  
9             https://en.wikipedia.org/wiki/Galadriel  
10  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck  
11                                                NaN

Sometimes it's nice to know how a particular row got added to the resulting DataFrame.  Using ```indicator=True```
allows us to examine this:

In [100]:
df.merge(urls_complete,how='outer',indicator=True)

name    race  magic  aggression  stealth  \
0     Gandalf    Maia   10.0         7.0      8.0   
1       Gimli   Dwarf    1.0        10.0      2.0   
2       Frodo  Hobbit    4.0         2.0      5.0   
3     Legolas     Elf    6.0         5.0     10.0   
4       Bilbo  Hobbit    4.0         1.0      5.0   
5         Sam  Hobbit    2.0         6.0      4.0   
6      Pippin  Hobbit    0.0         3.0      5.0   
7     Boromir     Man    0.0         8.0      3.0   
8     Aragorn     Man    2.0         7.0      9.0   
9   Galadriel     Elf    9.0         2.0     10.0   
10   Meriadoc  Hobbit    0.0         4.0      6.0   
11       Lily  Hobbit    NaN         NaN      NaN   
12  Treebeard     NaN    NaN         NaN      NaN   
13     Elrond     NaN    NaN         NaN      NaN   

                                                  url      _merge  
0               https://en.wikipedia.org/wiki/Gandalf        both  
1   https://en.wikipedia.org/wiki/Gimli_(Middle-ea...        both  
2         https://en.wikipedia.org/wiki/Frodo_Baggins        both  
3               https://en.wikipedia.org/wiki/Legolas        both  
4         https://en.wikipedia.org/wiki/Bilbo_Baggins        both  
5        https://en.wikipedia.org/wiki/Samwise_Gamgee        both  
6         https://en.wikipedia.org/wiki/Peregrin_Took        both  
7               https://en.wikipedia.org/wiki/Boromir        both  
8               https://en.wikipedia.org/wiki/Aragorn        both  
9             https://en.wikipedia.org/wiki/Galadriel        both  
10  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck        both  
11                                                NaN        both  
12            https://en.wikipedia.org/wiki/Treebeard  right_only  
13               https://en.wikipedia.org/wiki/Elrond  right_only

You'll note that we used the ```merge``` function from the DataFrame and passed in the other DataFrame as an argument.
You can also call the ```merge``` function from pandas directly and pass it the two DataFrames you are merging:

![pivot 1](http://www.datasciencemadesimple.com/wp-content/uploads/2017/09/join-or-merge-in-python-pandas-1.png "pivots")

In [94]:
pd.merge(df,urls_complete,how='outer',indicator=True)

name    race  magic  aggression  stealth  \
0     Gandalf    Maia   10.0         7.0      8.0   
1       Gimli   Dwarf    1.0        10.0      2.0   
2       Frodo  Hobbit    4.0         2.0      5.0   
3     Legolas     Elf    6.0         5.0     10.0   
4       Bilbo  Hobbit    4.0         1.0      5.0   
5         Sam  Hobbit    2.0         6.0      4.0   
6      Pippin  Hobbit    0.0         3.0      5.0   
7     Boromir     Man    0.0         8.0      3.0   
8     Aragorn     Man    2.0         7.0      9.0   
9   Galadriel     Elf    9.0         2.0     10.0   
10   Meriadoc  Hobbit    0.0         4.0      6.0   
11       Lily  Hobbit    NaN         NaN      NaN   
12  Treebeard     NaN    NaN         NaN      NaN   
13     Elrond     NaN    NaN         NaN      NaN   

                                                  url      _merge  
0               https://en.wikipedia.org/wiki/Gandalf        both  
1   https://en.wikipedia.org/wiki/Gimli_(Middle-ea...        both  
2         https://en.wikipedia.org/wiki/Frodo_Baggins        both  
3               https://en.wikipedia.org/wiki/Legolas        both  
4         https://en.wikipedia.org/wiki/Bilbo_Baggins        both  
5        https://en.wikipedia.org/wiki/Samwise_Gamgee        both  
6         https://en.wikipedia.org/wiki/Peregrin_Took        both  
7               https://en.wikipedia.org/wiki/Boromir        both  
8               https://en.wikipedia.org/wiki/Aragorn        both  
9             https://en.wikipedia.org/wiki/Galadriel        both  
10  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck        both  
11                                                NaN        both  
12            https://en.wikipedia.org/wiki/Treebeard  right_only  
13               https://en.wikipedia.org/wiki/Elrond  right_only

## <font color="magenta">Q8: (4 points) Join the ```url_df``` DataFrame (that contains aliases) to the ```df``` DataFrame using an appropriate merge

In [101]:
df.merge(url_df, how='inner')

name    race  magic  aggression  stealth  \
0     Gimli   Dwarf    1.0        10.0      2.0   
1     Frodo  Hobbit    4.0         2.0      5.0   
2   Legolas     Elf    6.0         5.0     10.0   
3    Pippin  Hobbit    0.0         3.0      5.0   
4  Meriadoc  Hobbit    0.0         4.0      6.0   

                                                 url  \
0  https://en.wikipedia.org/wiki/Gimli_(Middle-ea...   
1        https://en.wikipedia.org/wiki/Frodo_Baggins   
2              https://en.wikipedia.org/wiki/Legolas   
3        https://en.wikipedia.org/wiki/Peregrin_Took   
4  https://en.wikipedia.org/wiki/Meriadoc_Brandybuck   

                                             aliases  
0  Elf-friend Lockbearer Lord of the Glittering C...  
1                                      Mr. Underhill  
2        Greenleaf, (Legolas translatedinto English)  
3                                             Pippin  
4                                               None

# Aggregation and Grouping

## Some more basic pandas functionality
One of the nice things about pandas is that it simplifies many common operations on datasets.  Let's load yet another LOTR dataset, 
this time using StringIO.  StringIO allows us to create a string that's then available as a file!

Why would you want to do this?  Sometimes it's easier to just paste data right into a Jupyter notebook (or python script)
than it is to create another CSV file.


In [95]:
from io import StringIO

LOTRDATA=StringIO("""name,race,gender,magic,aggression,stealth
Gandalf,Maia,Male,10.0,7.0,8.0
Gimli,Dwarf,None,1.0,10.0,2.0
Frodo,Hobbit,Male,4.0,2.0,5.0
Legolas,Elf,Male,6.0,5.0,10.0
Bilbo,Hobbit,Male,4.0,1.0,5.0
Sam,Hobbit,Male,2.0,6.0,4.0
Pippin,Hobbit,Male,0.0,3.0,5.0
Boromir,Human,Male,0.0,8.0,3.0
Aragorn,Human,Male,2.0,7.0,9.0
Galadriel,Elf,Female,9.0,2.0,10.0
Lily,Hobbit,Female,,,
Meriadoc,Hobbit,Male,,4.0,6.0
Melian,Maia,Female,10.0,5.0,9.0
Idril,Elf,Female,8.0,,8.0
""")

lotr = pd.read_csv(LOTRDATA, index_col=None)
lotr

name    race  gender  magic  aggression  stealth
0     Gandalf    Maia    Male   10.0         7.0      8.0
1       Gimli   Dwarf    None    1.0        10.0      2.0
2       Frodo  Hobbit    Male    4.0         2.0      5.0
3     Legolas     Elf    Male    6.0         5.0     10.0
4       Bilbo  Hobbit    Male    4.0         1.0      5.0
5         Sam  Hobbit    Male    2.0         6.0      4.0
6      Pippin  Hobbit    Male    0.0         3.0      5.0
7     Boromir   Human    Male    0.0         8.0      3.0
8     Aragorn   Human    Male    2.0         7.0      9.0
9   Galadriel     Elf  Female    9.0         2.0     10.0
10       Lily  Hobbit  Female    NaN         NaN      NaN
11   Meriadoc  Hobbit    Male    NaN         4.0      6.0
12     Melian    Maia  Female   10.0         5.0      9.0
13      Idril     Elf  Female    8.0         NaN      8.0

Now that we have a DataFrame, we can get some basic statistics about it using the ```describe()``` function.
Note that ```describe()``` only returns values for numeric columns.  Note too that it returns another DataFrame

In [96]:
lotr.describe()

magic  aggression    stealth
count  12.000000   12.000000  13.000000
mean    4.666667    5.000000   6.461538
std     3.821788    2.730301   2.696151
min     0.000000    1.000000   2.000000
25%     1.750000    2.750000   5.000000
50%     4.000000    5.000000   6.000000
75%     8.250000    7.000000   9.000000
max    10.000000   10.000000  10.000000

In [97]:
type(lotr.describe())

pandas.core.frame.DataFrame

## Pivots and Pivot Tables
The following cells are based on: 
http://nikgrozev.com/2015/07/01/reshaping-in-pandas-pivot-pivot-table-stack-and-unstack-explained-with-pictures/ , which is
one of the best guides to pivots, pivot tables, stacking and unstacking that I've encountered.

For demonstration purposes, let's create the same DataFrame that Nikolay Grozev uses in his tutorial:

In [98]:
from collections import OrderedDict
from pandas import DataFrame

table = OrderedDict((
    ("Item", ['Item0', 'Item0', 'Item1', 'Item1']),
    ('CType',['Gold', 'Bronze', 'Gold', 'Silver']),
    ('USD',  ['1$', '2$', '3$', '4$']),
    ('EU',   ['1€', '2€', '3€', '4€'])
))
metal = DataFrame(table)
metal

# create the table

Item   CType USD  EU
0  Item0    Gold  1$  1€
1  Item0  Bronze  2$  2€
2  Item1    Gold  3$  3€
3  Item1  Silver  4$  4€

In [99]:
# make a table of items (rows) and costs (USD) 
# for each in gold and bronze
metal.pivot(index='Item',columns='CType',values='USD')

CType Bronze Gold Silver
Item                    
Item0     2$   1$    NaN
Item1    NaN   3$     4$

See the image below... we're telling Pandas to take the table above, create a row for every item. 
This is done by setting index to Item (the column in the original table that contains item names)
We then are telling pandas we want to create a column for every unique element in the
original CType column.  And finally, we want the value in the cells to be the value from the USD
column in the original table.

![pivot 1](http://nikgrozev.com/images/blog/Reshaping%20in%20Pandas%20-%20Pivot%20Pivot-Table%20Stack%20and%20Unstack%20explained%20with%20Pictures/pivoting_simple1.png "pivots")

## <font color="magenta">Q9: (2 points) Using pivot and the dataframe below make a table with each person as the index, their purchased items as columns, and the price they paid as the values.</font>


In [102]:
table1 = OrderedDict((
    ("Person", ['Scott', 'Julie', 'Shiyan', 'Julie','Scott','Julie', 'Shiyan', 'Julie','Scott', 'Julie']),
    ('Item Purchased',['Kit-Kat','Mango', 'Twix', 'M&Ms', 'Ferrero Rocher','Apple', 'Watermelon', 'Pineapple', 'Snickers','Pear']),
    ('Price Paid',  ['$1','$2', '$1.5', '$2', '$1.5','$1','$1','$2', '$1.5', '$2']),
    ))
fav = DataFrame(table1)
fav

Person  Item Purchased Price Paid
0   Scott         Kit-Kat         $1
1   Julie           Mango         $2
2  Shiyan            Twix       $1.5
3   Julie            M&Ms         $2
4   Scott  Ferrero Rocher       $1.5
5   Julie           Apple         $1
6  Shiyan      Watermelon         $1
7   Julie       Pineapple         $2
8   Scott        Snickers       $1.5
9   Julie            Pear         $2

In [103]:
fav.pivot(index='Person',columns='Item Purchased',values='Price Paid')

Item Purchased Apple Ferrero Rocher Kit-Kat M&Ms Mango Pear Pineapple  \
Person                                                                  
Julie             $1            NaN     NaN   $2    $2   $2        $2   
Scott            NaN           $1.5      $1  NaN   NaN  NaN       NaN   
Shiyan           NaN            NaN     NaN  NaN   NaN  NaN       NaN   

Item Purchased Snickers  Twix Watermelon  
Person                                    
Julie               NaN   NaN        NaN  
Scott              $1.5   NaN        NaN  
Shiyan              NaN  $1.5         $1

In [ ]:
# end of exercise

In [104]:
p = metal.pivot(index='Item',columns='CType',values='USD')  # same as above, I'm just renaming the pivot as "p"

Access the USD cost of Item0 for Gold customers...

First we find the row for Item0/Gold and then we select the USD column and pull out the value

In [105]:
result = metal[((metal.Item == 'Item0') & (metal.CType == 'Gold'))].USD

In [106]:
type(result)

pandas.core.series.Series

In [107]:
result

0    1$
Name: USD, dtype: object

Do the same thing on pivoted table. Here we pull out the row for Item0, grab the Gold column and print the value

In [108]:
p[p.index == 'Item0'].Gold.values

array(['1$'], dtype=object)

Now pivot by multiple columns, I want USD and EU prices. It returns a hierarchical index.

In [109]:
metal.pivot(index='Item',columns='CType')

USD                 EU            
CType Bronze Gold Silver Bronze Gold Silver
Item                                       
Item0     2$   1$    NaN     2€   1€    NaN
Item1    NaN   3$     4$    NaN   3€     4€

Access the USD cost of Item0 for Gold customers

In [110]:
p = metal.pivot(index='Item',columns='CType')
p.USD[p.USD.index == 'Item0'].Gold.values

array(['1$'], dtype=object)

# What happens if there is a collision? 
See the problem?  There are two Item0/Golds:


![pivot 2](http://nikgrozev.com/images/blog/Reshaping%20in%20Pandas%20-%20Pivot%20Pivot-Table%20Stack%20and%20Unstack%20explained%20with%20Pictures/pivoting_simple_error.png "pivots")

Let's set up another DataFrame to demonstrate this:

In [111]:
table = OrderedDict((
    ("Item", ['Item0', 'Item0', 'Item0', 'Item1']),
    ('CType',['Gold', 'Bronze', 'Gold', 'Silver']),
    ('USD',  ['1$', '2$', '3$', '4$']),
    ('EU',   ['1€', '2€', '3€', '4€'])
))
metal = DataFrame(table)


In [112]:
metal.head()

Item   CType USD  EU
0  Item0    Gold  1$  1€
1  Item0  Bronze  2$  2€
2  Item0    Gold  3$  3€
3  Item1  Silver  4$  4€

The next cell will generate an error:

In [113]:
p = metal.pivot(index='Item', columns='CType', values='USD')
# will return an error

ValueError: Index contains duplicate entries, cannot reshape

## pivot_tables is your friend
![pivot 1](http://nikgrozev.com/images/blog/Reshaping%20in%20Pandas%20-%20Pivot%20Pivot-Table%20Stack%20and%20Unstack%20explained%20with%20Pictures/pivoting_table_simple1.png "pivots")

Let's create yet another DataFrame to play with:

In [114]:
table = OrderedDict((
    ("Item", ['Item0', 'Item0', 'Item0', 'Item1']),
    ('CType',['Gold', 'Bronze', 'Gold', 'Silver']),
    ('USD',  [1, 2, 3, 4]),
    ('EU',   [1.1, 2.2, 3.3, 4.4])
))
metal = DataFrame(table)
metal

Item   CType  USD   EU
0  Item0    Gold    1  1.1
1  Item0  Bronze    2  2.2
2  Item0    Gold    3  3.3
3  Item1  Silver    4  4.4

pivot_table is  a bit different than pivot... It's the same with the first part
index, columns, values remain the same as before BUT we added a rule (aggfunc)
that says: whey you hit a conflict, the way to resolve it is X (in this case
x is the "mean"... so find the mean of the two numbers)

In [115]:
p = metal.pivot_table(index='Item',columns='CType',values='USD',aggfunc=np.mean)
p

CType  Bronze  Gold  Silver
Item                       
Item0     2.0   2.0     NaN
Item1     NaN   NaN     4.0

You could have also resolved the conflict in other ways.  Here we tell it to take the "min":

In [116]:
p = metal.pivot_table(index='Item',columns='CType',values='USD',aggfunc=np.min)
p

CType  Bronze  Gold  Silver
Item                       
Item0     2.0   1.0     NaN
Item1     NaN   NaN     4.0

## <font color="magenta">Q10: (4 points) Using pivot_table and the dataframe below make a table with each person as the index, their purchased items as columns, and the sum of the price they paid as the values (sometimes people bought more than one of the same item).</font>


In [117]:
table1 = OrderedDict((
    ("Person", ['Scott', 'Julie', 'Shiyan', 'Julie',
                'Scott','Julie', 'Shiyan', 'Julie','Scott', 
                'Julie','Scott', 'Julie', 'Shiyan', 'Julie',
                'Scott','Julie',]),
    ('Item Purchased',['Kit-Kat','Mango', 'Twix', 'M&Ms', 'Ferrero Rocher',
                       'Apple', 'Watermelon', 'Pineapple', 'Snickers','Pear','Kit-Kat','Mango', 'Twix', 'M&Ms', 'Ferrero Rocher',
                       'Apple']),
    ('Price Paid (USD)',  [1,2, 1.5, 2, 1.5,1,1,2, 1.5, 2,
                    1,2, 1.5, 2, 1.5,1]),
    ))
fav = DataFrame(table1)
fav

Person  Item Purchased  Price Paid (USD)
0    Scott         Kit-Kat               1.0
1    Julie           Mango               2.0
2   Shiyan            Twix               1.5
3    Julie            M&Ms               2.0
4    Scott  Ferrero Rocher               1.5
5    Julie           Apple               1.0
6   Shiyan      Watermelon               1.0
7    Julie       Pineapple               2.0
8    Scott        Snickers               1.5
9    Julie            Pear               2.0
10   Scott         Kit-Kat               1.0
11   Julie           Mango               2.0
12  Shiyan            Twix               1.5
13   Julie            M&Ms               2.0
14   Scott  Ferrero Rocher               1.5
15   Julie           Apple               1.0

In [118]:
fav.pivot_table(index='Person',columns='Item Purchased',values='Price Paid (USD)',aggfunc=sum)

Item Purchased  Apple  Ferrero Rocher  Kit-Kat  M&Ms  Mango  Pear  Pineapple  \
Person                                                                         
Julie             2.0             NaN      NaN   4.0    4.0   2.0        2.0   
Scott             NaN             3.0      2.0   NaN    NaN   NaN        NaN   
Shiyan            NaN             NaN      NaN   NaN    NaN   NaN        NaN   

Item Purchased  Snickers  Twix  Watermelon  
Person                                      
Julie                NaN   NaN         NaN  
Scott                1.5   NaN         NaN  
Shiyan               NaN   3.0         1.0

Pivots are a specific form of stack/unstack (remember those?)

![pivot 1](http://nikgrozev.com/images/blog/Reshaping%20in%20Pandas%20-%20Pivot%20Pivot-Table%20Stack%20and%20Unstack%20explained%20with%20Pictures/stack-unstack1.png "pivots")

## A worked example

In [119]:
# to start let's make a fake dataset: sales of fruit across US states.
# Don't worry about the details here, but basically we'll pretend
# this string is a CSV file and use the standard loading ops
from io import StringIO

TESTDATA=StringIO("""State,Retailer,Fruit,Sales
MI,Walmart,Apple,100
MI,Wholefoods,Apple,150
MI,Kroger,Orange,180
CA,Walmart,Apple,220
CA,Wholefoods,Apple,180
CA,Safeway,Apple,220
CA,Safeway,Orange,110
NY,Walmart,Apple,90
NY,Walmart,Orange,80
NY,Wholefoods,Orange,120
""")

fruit = pd.read_csv(TESTDATA, index_col=None)
fruit

State    Retailer   Fruit  Sales
0    MI     Walmart   Apple    100
1    MI  Wholefoods   Apple    150
2    MI      Kroger  Orange    180
3    CA     Walmart   Apple    220
4    CA  Wholefoods   Apple    180
5    CA     Safeway   Apple    220
6    CA     Safeway  Orange    110
7    NY     Walmart   Apple     90
8    NY     Walmart  Orange     80
9    NY  Wholefoods  Orange    120

## (a) What is the total sales for each state?
This requires us to group by state, and aggregate sales by taking the sum.

The easiest way of doing this if to use `groupby`

If you execute groupby on the dataframe what you'll get back is an object called DataFrameGroupBy

In [120]:
fruit.groupby('State')

On its own it's a bit useless... it just keeps track of which rows should go into each "pile" (where pile here means a unique group for each state)

If we ask this object to describe itself, you can see what is inside notice that it threw away all the other columns because they were not numerical.  Only "Sales" which is a number, was kept

In [121]:
fruit.groupby('State').describe()

Sales                                                          
      count        mean        std    min    25%    50%    75%    max
State                                                                
CA      4.0  182.500000  51.881275  110.0  162.5  200.0  220.0  220.0
MI      3.0  143.333333  40.414519  100.0  125.0  150.0  165.0  180.0
NY      3.0   96.666667  20.816660   80.0   85.0   90.0  105.0  120.0

Now, if we had another numerical column, let's call it "Sales2," that column would also be kept.  Let's make a fruit2 DataFrame so you can see that:

In [122]:
from io import StringIO

TESTDATA=StringIO("""State,Retailer,Fruit,Sales,Sales2
MI,Walmart,Apple,100,10
MI,Wholefoods,Apple,150,20
MI,Kroger,Orange,180,30
CA,Walmart,Apple,220,20
CA,Wholefoods,Apple,180,40
CA,Safeway,Apple,220,30
CA,Safeway,Orange,110,20
NY,Walmart,Apple,90,40
NY,Walmart,Orange,80,20
NY,Wholefoods,Orange,120,60
""")

fruit2 = pd.read_csv(TESTDATA, index_col=None)
fruit2

State    Retailer   Fruit  Sales  Sales2
0    MI     Walmart   Apple    100      10
1    MI  Wholefoods   Apple    150      20
2    MI      Kroger  Orange    180      30
3    CA     Walmart   Apple    220      20
4    CA  Wholefoods   Apple    180      40
5    CA     Safeway   Apple    220      30
6    CA     Safeway  Orange    110      20
7    NY     Walmart   Apple     90      40
8    NY     Walmart  Orange     80      20
9    NY  Wholefoods  Orange    120      60

In [123]:
fruit2.groupby("State").describe()

Sales                                                           Sales2  \
      count        mean        std    min    25%    50%    75%    max  count   
State                                                                          
CA      4.0  182.500000  51.881275  110.0  162.5  200.0  220.0  220.0    4.0   
MI      3.0  143.333333  40.414519  100.0  125.0  150.0  165.0  180.0    3.0   
NY      3.0   96.666667  20.816660   80.0   85.0   90.0  105.0  120.0    3.0   

                                                      
       mean        std   min   25%   50%   75%   max  
State                                                 
CA     27.5   9.574271  20.0  20.0  25.0  32.5  40.0  
MI     20.0  10.000000  10.0  15.0  20.0  25.0  30.0  
NY     40.0  20.000000  20.0  30.0  40.0  50.0  60.0

To actually make use of the groupby, we need to tell pandas what to use to measure what's in each group. In other words, I've created a pile for California, a pile for Michigan, and a pile for New York.  I want a number to what's *inside* each pile.  I could ask for the "size" (so how many rows are in each pile), or I could calculate some mathematical function.  For example, if I wanted to know the total sales, I would call "sum."  What happens is pandas goes through every pile, looks at every "row" inside that pile and, for all numerical properties, calculated something.  In this case it's sum... it adds up everything.  So in our original table we had three items for Michigan (Walmart, Kroger, Wholefoods).  This is our Michigan pile.  We then look at numerical properties for Walmart, Kroger, and Wholefoods. In this case Sales.  Because we are using sum() that means add the sales of each.

In [124]:
# What are the total sales for each state?
fruit.groupby('State').sum()  # instead of size()

Sales
State       
CA       730
MI       430
NY       290

What just happend? A couple of things:
- `groupby()` got first executed on `df`, returning an `DataFrameGroupBy` object. This object itself is useless unless coupled with an aggregation function, such as `sum()`, `mean()`, `max()`, `apply()`. 
- Then, `sum()` got executed on the `DataFrameGroupBy` object, generating the `DataFrame` object you see above. Notice how the table looks different than the original DataFrame `df`? Here are the differences:
  - The `State` column now becomes the index of the DataFrame. The string "State" is the name of the index. Notice how the index name is displayed on a lower level than column names.
  - Since we performed a `groupby` operation by `State`, so only the unique values of `State` are kept as index.
  - Among the other columns, Retailer, Fruit, and Sales, only Sales is kept in the result table. This is because the aggregation function `sum()` only knows how to aggregate numerical values. And only Sales is a numerical column. The other columns are hence dropped.

## <font color="magenta">Q11: (2 points) Using groupby, which Retailer had the highest total Sales and how much was that:

In [125]:
fruit.groupby('Retailer').sum().sort_values('Sales').tail(2)

Sales
Retailer         
Wholefoods    450
Walmart       490

Walmart had the highest total Sales, and that was 490.

## (b) What is the total sales for each state for each fruit?
This requires us to perform `groupby` on two columns. So, we provide a list of column names to the `groupby` function.

Don't forget that an aggregation function needs to follow the `groupby` function in order to generate results.

In [126]:
# What is the total sales for each state for each fruit?
fruit.groupby(['State','Fruit']).sum()

Sales
State Fruit        
CA    Apple     620
      Orange    110
MI    Apple     250
      Orange    180
NY    Apple      90
      Orange    200

How is this DataFrame different from the previous one?

The biggest different is that this DataFrame has what is called a `MultiIndex` (or hierarchical index), as opposed to a simple index. In this table, the left two "columns" are not columns but actually part of the `MultiIndex`, and the `Sales` is the single real "column" in the DataFrame. (Running out of terminologies here...)

The hierarchical index can be organized in an alternative way if we swapped the order of State and Fruit.

In [127]:
fruit.groupby(['Fruit','State']).sum()

Sales
Fruit  State       
Apple  CA       620
       MI       250
       NY        90
Orange CA       110
       MI       180
       NY       200

## (c) Which state has the maximum total sales?
This question is not asking about the maximum value, but rather which state holds that maximum. There are multiple ways to do it. A principled way is to use `idxmax`.

In [128]:
# Which state has the maximum total sales?
fruitSalesByState = fruit.groupby('State').sum()
print(fruitSalesByState)
max_state = fruitSalesByState['Sales'].idxmax()
print("The state with the maximum sales is: ",max_state)

       Sales
State       
CA       730
MI       430
NY       290
The state with the maximum sales is:  CA




What if I want to display the maximum value alongside the state? Well, we can use that returned label to _select_ the corresponding row from the original DataFrame.

In [129]:
# the deets
fruitSalesByState.loc['CA']

Sales    730
Name: CA, dtype: int64

A less efficient but more intuitive way of doing the same thing:

In [130]:
fruitSalesByState.sort_values('Sales',ascending=False).iloc[0]

Sales    730
Name: CA, dtype: int64

### Which state has the maximum total sales for apples?</font>

In [131]:
# Which state has the maximum total sales for apples?
# give me apple sellers
apples = fruit[fruit.Fruit == 'Apple']
apples

State    Retailer  Fruit  Sales
0    MI     Walmart  Apple    100
1    MI  Wholefoods  Apple    150
3    CA     Walmart  Apple    220
4    CA  Wholefoods  Apple    180
5    CA     Safeway  Apple    220
7    NY     Walmart  Apple     90

In [132]:
# aggr. by state
applesByState = apples.groupby('State').sum()
applesByState

Sales
State       
CA       620
MI       250
NY        90

In [133]:
applesByState.Sales

State
CA    620
MI    250
NY     90
Name: Sales, dtype: int64

In [134]:
applesByState.Sales.idxmax()

'CA'

In [135]:
applesByState.loc[applesByState.Sales.idxmax()]

Sales    620
Name: CA, dtype: int64

In the above command, `.loc[]` looks up the index label and returns that row.



## Applying what we learned to the LOTR data

What are the average values for magic, aggression, and stealth for each race?

In [136]:
from io import StringIO

LOTRDATA=StringIO("""name,race,gender,magic,aggression,stealth
Gandalf,Maia,Male,10.0,7.0,8.0
Gimli,Dwarf,None,1.0,10.0,2.0
Frodo,Hobbit,Male,4.0,2.0,5.0
Legolas,Elf,Male,6.0,5.0,10.0
Bilbo,Hobbit,Male,4.0,1.0,5.0
Sam,Hobbit,Male,2.0,6.0,4.0
Pippin,Hobbit,Male,0.0,3.0,5.0
Boromir,Human,Male,0.0,8.0,3.0
Aragorn,Human,Male,2.0,7.0,9.0
Galadriel,Elf,Female,9.0,2.0,10.0
Lily,Hobbit,Female,,,
Meriadoc,Hobbit,Male,,4.0,6.0
Melian,Maia,Female,10.0,5.0,9.0
Idril,Elf,Female,8.0,,8.0
""")

lotr = pd.read_csv(LOTRDATA, index_col=None)

In [137]:
lotr.head()

name    race gender  magic  aggression  stealth
0  Gandalf    Maia   Male   10.0         7.0      8.0
1    Gimli   Dwarf   None    1.0        10.0      2.0
2    Frodo  Hobbit   Male    4.0         2.0      5.0
3  Legolas     Elf   Male    6.0         5.0     10.0
4    Bilbo  Hobbit   Male    4.0         1.0      5.0

## <font color="magenta">Q12: (4 points) Create a pivot table showing the maximum values of magic using gender for columns and race for rows.  

In [139]:
lotr.pivot_table(index='race',columns='gender',values='magic',aggfunc=max)

gender  Female  Male  None
race                      
Dwarf      NaN   NaN   1.0
Elf        9.0   6.0   NaN
Hobbit     NaN   4.0   NaN
Human      NaN   2.0   NaN
Maia      10.0  10.0   NaN

# <font color="magenta">END OF NOTEBOOK</font>
## Remember to submit this file in HTML and IPYNB formats via Canvas.